src/ff1/alloc.rs

```diff
@@ -121,86 +168,237 @@ impl BinaryNumeralString {
     /// Creates a BinaryNumeralString from a byte slice, with each byte
     /// interpreted in little-endian bit order.
     pub fn from_bytes_le(s: &[u8]) -> Self {
-        let mut data = Vec::with_capacity(s.len() * 8);
-        for n in s {
-            let mut tmp = *n;
-            for _ in 0..8 {
-                data.push(tmp & 1);
-                tmp >>= 1;
-            }
-        }
-        BinaryNumeralString(data)
+        BinaryNumeralString(s.to_vec())
     }
 
     /// Returns a Vec<u8>, with each byte written from the BinaryNumeralString
     /// in little-endian bit order.
     pub fn to_bytes_le(&self) -> Vec<u8> {
-        // We should always have a multiple of eight bits
-        assert_eq!((self.0.len() + 7) / 8, self.0.len() / 8);
-        let mut data = Vec::with_capacity(self.0.len() / 8);
-        let mut acc = 0;
-        let mut shift = 0;
-        for n in &self.0 {
-            acc += n << shift;
-            shift += 1;
-            if shift == 8 {
-                data.push(acc);
-                acc = 0;
-                shift = 0;
-            }
-        }
-        data
+        self.0.to_vec()
     }
 }
 
 impl NumeralString for BinaryNumeralString {
-    type Num = BigUint;
+    type Ops = BinaryOps;
 
     fn is_valid(&self, radix: u32) -> bool {
-        self.0.iter().all(|n| (u32::from(*n) < radix))
+        // This struct is valid for radix 2 by construction.
+        radix == 2
     }
 
     fn numeral_count(&self) -> usize {
-        self.0.len()
+        self.0.len() * 8
     }
 
-    fn split(&self, u: usize) -> (Self, Self) {
-        let mut front = self.0.clone();
-        let back = front.split_off(u);
-        (BinaryNumeralString(front), BinaryNumeralString(back))
+    fn split(&self) -> (Self::Ops, Self::Ops) {
+        let n = self.numeral_count();
+        let u = n / 2;
+        let v = n - u;
+        let a_end = (u + 7) / 8;
+        let b_start = u / 8;
+
+        // FF1 processes the two halves of a numeral string as big-endian integers in the
+        // given radix, via the `NUM_radix()` operation. We are operating on binary data
+        // with a radix of 2, which means the "bit string" is interpreted as big endian.
+        //
+        // However, `BinaryNumeralString::from_bytes_le` uses little-endian bit order when
+        // parsing a byte encoding into a bit string (which indeed it should, otherwise
+        // the byte encoding would be mixed-endian which no one should have to suffer).
```

Comment:
Just to make sure that we've got the usage of this correct in zcashd and the Zcash protocol spec:

In the spec, [§ 5.4.4 Pseudo Random Permutations](https://zips.z.cash/protocol/protocol.pdf#concreteprps), the plaintext input to FF1-AES256 is a diversifier index which is a sequence of 88 bits. Similarly a diversifier is a sequence of 88 bits. The conversion of a diversifier to a byte sequence is done as $\mathsf{LEBS2OSP}_{88}(\mathsf{d})$ in [§ 5.4.1.6 Diversify<sup>Sapling</sup> and DiversifyHash<sup>Orchard</sup> Hash Functions](https://zips.z.cash/protocol/protocol.pdf#concretediversifyhash), which uses little-endian bit order.

In practice we represent diversifiers and diversifier indices as byte arrays in the C++ and Rust code (e.g. [here](https://github.com/zcash/librustzcash/blob/fb49cd412f02b65def53a096f536c2d4aeda68f4/zcash_primitives/src/sapling/keys.rs#L275) for `sapling::keys::Diversifier` in librustzcash), and those are to be interpreted as being in little-endian bit order, so that there is no bit reversal in the $\mathsf{DiversifyHash^*}$ implementations.

So it is correct, for instance, to use `BinaryNumeralString::from_bytes_le` in [zip32::sapling::DiversifierKey::try_diversifier_internal](https://github.com/zcash/librustzcash/blob/fb49cd412f02b65def53a096f536c2d4aeda68f4/zcash_primitives/src/zip32/sapling.rs#L178) and other places where we do FF1 encryption/decryption.

Reply:
(No code change needed.)